core/channel-upgrade.go

```diff
@@ -0,0 +1,718 @@
+
+	if _, err := chain.SendMsgs([]sdk.Msg{msg}); err != nil {
+		logger.Error("failed to send MsgChannelUpgradeCancel", err)
+		return err
+	} else {
+		logger.Info("successfully cancelled the channel upgrade")
+	}
+
+	return nil
+}
+
+func NewUpgradeState(chanState chantypes.State, upgradeExists bool) (UpgradeState, error) {
+	switch chanState {
+	case chantypes.OPEN:
+		if upgradeExists {
+			return UPGRADE_STATE_INIT, nil
+		} else {
+			return UPGRADE_STATE_UNINIT, nil
+		}
+	case chantypes.FLUSHING:
+		return UPGRADE_STATE_FLUSHING, nil
+	case chantypes.FLUSHCOMPLETE:
+		return UPGRADE_STATE_FLUSHCOMPLETE, nil
+	default:
+		return 0, fmt.Errorf("channel not opened yet: state=%s", chanState)
+	}
+}
+
+func upgradeChannelStep(src, dst *ProvableChain, targetSrcState, targetDstState UpgradeState, firstCall bool) (*RelayMsgs, error) {
+	logger := GetChannelPairLogger(src, dst)
+
+	out := NewRelayMsgs()
+	if err := validatePaths(src, dst); err != nil {
+		logger.Error("failed to validate paths", err)
+		return nil, err
+	}
+
+	// First, update the light clients to the latest header and return the header
+	sh, err := NewSyncHeaders(src, dst)
+	if err != nil {
+		logger.Error("failed to create SyncHeaders", err)
+		return nil, err
+	}
+
+	// Query a number of things all at once
+	var srcUpdateHeaders, dstUpdateHeaders []Header
+	if err := retry.Do(func() error {
+		srcUpdateHeaders, dstUpdateHeaders, err = sh.SetupBothHeadersForUpdate(src, dst)
+		return err
+	}, rtyAtt, rtyDel, rtyErr, retry.OnRetry(func(uint, error) {
+		if err := sh.Updates(src, dst); err != nil {
+			panic(err)
+		}
+	})); err != nil {
+		logger.Error("failed to set up headers for LC update on both chains", err)
+		return nil, err
+	}
+
+	// prepare query contexts
+	var srcCtxFinalized, dstCtxFinalized, srcCtxLatest, dstCtxLatest QueryContext
+	if srcCtxFinalized, err = getQueryContext(src, sh, true); err != nil {
+		logger.Error("failed to get query the context based on the src chain's latest finalized height", err)
+		return nil, err
+	}
+	if dstCtxFinalized, err = getQueryContext(dst, sh, true); err != nil {
+		logger.Error("failed to get query the context based on the dst chain's latest finalized height", err)
+		return nil, err
+	}
+	if srcCtxLatest, err = getQueryContext(src, sh, false); err != nil {
+		logger.Error("failed to get query the context based on the src chain's latest height", err)
+		return nil, err
+	}
+	if dstCtxLatest, err = getQueryContext(dst, sh, false); err != nil {
+		logger.Error("failed to get query the context based on the dst chain's latest height", err)
+		return nil, err
+	}
+
+	// query finalized channels with proofs
+	srcChan, dstChan, err := QueryChannelPair(
+		srcCtxFinalized,
+		dstCtxFinalized,
+		src,
+		dst,
+		true,
+	)
+	if err != nil {
+		logger.Error("failed to query the channel pair with proofs", err)
+		return nil, err
+	} else if finalized, err := checkChannelFinality(src, dst, srcChan.Channel, dstChan.Channel); err != nil {
```

Comment:
Looks like `checkChannelFinality` only checks if the latest channel state matches the finalized one, but it does not check `upgrade_sequence` difference. Is this no problem?

Reply:
@bluele @bluele fixed!: https://github.com/hyperledger-labs/yui-relayer/pull/143/commits/a9f0af686251836f0f4a7889b7965f840fb31ff0